### DISCUSSION TITLE
FDS import snapshots for validation/benchmark

### DISCUSSION THREAD

**Discussion by fbraennstroem** [MaCFP/radi-db#3] [Ideas]
Hi all,

we still need a way to get snapshot data from the PMC-LBL to FDS to be able to
benchmark FDS as well. In addition, one would need to be able to activate only
the radiation calculation without any moment/species or energy equation.

## General
For the discussion, I would split it up into two parts:
1. read in data to FDS; I would call it IMPORT
2. add a flag to activate only radiation calculation; I would call it RAD_FLAG

There are certainly different options to get both running and I have 
the least experience with FDS, so I maybe just start the discussion with some
simple ideas for the different steps:

## IMPORT
a) Using &INIT line
- could be an option for a few cells or regions, but seems to be 
b) Add procedure to read csv in fortran
```x,y,z,field1```
- probably it is easiest to have one separate csv for each mesh and snapshot (which we already
  get out by fdsPOST.jl from FDS itself); so one would have also the same
  sorting order of the data
- maybe the position is not needed (at least for the FM Burner which gets the
  data from FDS)
- cell centered data are included
- even though having 
c) writing/Reading a restart file
- Format is not clear to me yet (I did not look)...
- Not sure, if this works smoothly for 128 meshes!?
- Maybe tricky, if not the standard fields are provided which are written
  usually by to the restart fields!?
d) define which binary FDS sf fields are reread without the process using
fdsPOST.jl; this would be completely within FDS

Questions / Comments:
- the mentioned csv files just load the internal field data; face/wall data are
  set to zero gradient!?

  
Is the &INIT  procedure here, `init.f90` line 429 and then slightly adjusted
again in 475?
```
! Over-ride default ambient conditions with user-prescribed INITializations

DO N=1,N_INIT
   IN => INITIALIZATION(N)
   IF (IN%NODE_ID/='null') CYCLE
   DO K=0,KBP1
      DO J=0,JBP1
         DO I=0,IBP1
            IF (M%XC(I) > IN%X1 .AND. M%XC(I) < IN%X2 .AND. &
                M%YC(J) > IN%Y1 .AND. M%YC(J) < IN%Y2 .AND. &
                M%ZC(K) > IN%Z1 .AND. M%ZC(K) < IN%Z2) THEN
               M%TMP(I,J,K)            = IN%TEMPERATURE
               M%RHO(I,J,K)            = IN%DENSITY
               M%RHOS(I,J,K)           = IN%DENSITY
               M%ZZ(I,J,K,1:N_TRACKED_SPECIES) = IN%MASS_FRACTION(1:N_TRACKED_SPECIES)
               M%ZZS(I,J,K,1:N_TRACKED_SPECIES) = IN%MASS_FRACTION(1:N_TRACKED_SPECIES)
               IF (IN%ADJUST_DENSITY)     M%RHO(I,J,K) = M%RHO(I,J,K)*M%P_0(K)/P_INF
               IF (IN%ADJUST_DENSITY)     M%RHOS(I,J,K) = M%RHOS(I,J,K)*M%P_0(K)/P_INF
               IF (IN%ADJUST_TEMPERATURE) M%TMP(I,J,K) = M%TMP(I,J,K)*M%P_0(K)/P_INF
            ENDIF
         ENDDO
      ENDDO
   ENDDO
ENDDO
```

## RAD_FLAG
a) no idea; in OpenFOAM I can just comment it out in the c++ code...
Questions / Comments:
- Might also be important as one might not load all needed fields to run a
  general simulation (e.g. openfoam would complain that there is a missing nut-field).
- maybe this is not needed, when we just run for one time step with a tiny time
  step size; so any momentum/energy interaction can be negelected (we could test
  this before in openfoam). But here we still could have the issue that we need
  to provide more field to be able to get a time step running than we can provide.

So these are just some ideas... what do you think?

Best regards
Fabian

**Comment by rmcdermo** [comment]
@fbraennstroem I am not sure I understand what we are trying to do.  The PMC-LBL gives KAPPA(I,J,K,BAND) and INTENSITY(I,J,K,BAND), yes?  Is the object to see if FDS RTE computes the correct radiative flux to a target given the correct radiation field?

To answer your other questions.

## Import
I think that INIT will work here.

## RAD_FLAG
I think that FREEZE_VELOCITY should work here.

**Reply by fbraennstroem** [reply]
Sorry, it is slightly different. The idea is to run FDS with the exact same snapshots-fields which were provided to the PMC-LBL. So FDS should get the field of the species and temperature start its own KAPPA/INTENSITY and radiative flux calculation. The calculation could then be done with different settings for each time step and compared with PMC-LBL.

**Reply by rmcdermo** [reply]
OK, thanks.  For this case, I think the thing to do is treat it like some of our other one-off verification cases, like the Shunn manufactured solution (see fds/Verification/Scalar_Analytical_Solution).  We have this (now poorly named) flag on the MISC line, PERIODIC_TEST.  We assign a new number for this case and then we can do a special initialization.  You can do this in your own branch of the code.  I can help set this up (after Tuesday next week).

**Reply by shostikk** [reply]
Actually, we want to import the input data for RTE solver, which means species concentrations + temperature (need density too?), and then let RTE solver calculate 
a) radiative properties (KAPPA, or several kappas, etc, depending on which method is in use),
b) intensitities and radiative heat fluxes.

Then we validate these calculations using the PMC-LBL predictions.

In one case, these snapshots were originally generated by FDS, so it may be easier to read in that original data. However, to be 100 % consistent with PMC-LBL solution, we may want to reload their snapshot. Maybe there were some changes in the data on the way.
In another cases, the snapshot was generated by OpenFOAM.

**Reply by rmcdermo** [reply]
Looking at the code.  Another option would be to expand the functionality of the CSVF namelist to include species and temperature (now is only works for velocity).  This would give the ability to read in 3D fields from ascii files.  What format are the data in currently?

**Reply by fbraennstroem** [reply]
Thanks, a simple option is to provide the data as csv with x,y,z,field for each mesh. This seems to have all what we need for reading in the data (except the other fields as mentioned).
I did run a small test case with 4x4x4 cells; the output seems to include one additional value in each direction.

Testcase
```
&HEAD CHID='box_single'/

&MESH IJK=4,4,4, XB=0,1,0,1,0,1/

&TIME T_END=1/

&DUMP DT_SLCF=0.1 ,DT_UVW=0.1/

&VENT MB='XMIN', SURF_ID='OPEN'/
&VENT MB='XMAX', SURF_ID='OPEN'/
&VENT MB='YMIN', SURF_ID='OPEN'/
&VENT MB='YMAX', SURF_ID='OPEN'/
&VENT MB='ZMAX', SURF_ID='OPEN'/

&INIT XB=0.25,0.75,0.25,0.75,0.25,0.75, TEMPERATURE=100/

&SLCF PBY=0.5, QUANTITY='TEMPERATURE'/
&SLCF PBY=0.5, QUANTITY='TEMPERATURE', CELL_CENTERED=T/

&TAIL/
```
Maybe we can talk on Wednesday or Thursday about it... have a nice weekend!

**Reply by fbraennstroem** [reply]
Hi, I just took a quick to the CSVF reading and writing.

**Comment by fab6** [comment]
Just a short update: Randy extended the CSVF option for temperature and species; I am testing it now. The initial generic test worked good; the step to the FM Burner was maybe too quick. It is still running since about 14 hours now...

**Comment by rmcdermo** [comment]
14 hours just for the initialization?  Something must be wrong.  How big are the files?

**Comment by fab6** [comment]
yes... each csv is ca. 330 kb; 64 of them. There is no error message. I am just simplifying a bit and give an update later today

**Comment by fab6** [comment]
I just testing temperature for now
![image](https://github.com/MaCFP/radi-db/assets/3047695/763c502c-0a36-4c38-904f-e64867c3b3bf)

primitives should be set for later species testing (as soon as temperature workds):
![image](https://github.com/MaCFP/radi-db/assets/3047695/a3cbc2b8-3dca-4c06-9292-6127d6082582)

**Comment by fab6** [comment]
Here it stops
![image](https://github.com/MaCFP/radi-db/assets/3047695/992da467-575a-49b1-97cc-4a2133a25e5c)

and section of temperature csv file:
![image](https://github.com/MaCFP/radi-db/assets/3047695/cbaa01ef-d5ef-4344-8d7b-f36fb64ed4b8)

and the naming for the temperature files:
![image](https://github.com/MaCFP/radi-db/assets/3047695/f67fdcce-f7ca-4704-b832-37ff4970ce66)

**Comment by rmcdermo** [comment]
Please send me the input file and csv files in a zip.  Thanks

**Comment by rmcdermo** [comment]
Fabian,
This seems to run fine for me in development (dv) and release versions.  Please try to see if your system has any issue with memory or something.

<img width="752" alt="Screenshot 2023-05-15 at 1 16 28 PM" src="https://github.com/MaCFP/radi-db/assets/4418497/7352234a-5f0e-4cb0-9391-f36d5476f452">

**Comment by rmcdermo** [comment]
BTW, only takes 9 seconds.

**Comment by fab6** [comment]
ha, cool... thanks, 9 seconds is indeed a bit faster than my run.
How much memory did you have for each core?

**Comment by rmcdermo** [comment]
We have 5 GB per core.

**Comment by fab6** [comment]
I think we have 2GB, I will see if I get 4GB...

**Comment by fab6** [comment]
![image](https://github.com/MaCFP/radi-db/assets/3047695/d35aa11a-e8dd-4a63-a449-449ddb6b0db1)
not working yet.. memory seems to be low

**Comment by fab6** [comment]
Not sure what this hydra is there for, but I think this is standard:
![image](https://github.com/MaCFP/radi-db/assets/3047695/34aa81b3-6822-4ccc-9e1b-5a5f42994e3e)

**Comment by rmcdermo** [comment]
Fabian,
Are you trying to use the `csvf` branch?  Maybe you missed my comment, but I merged all my updates to the main branch.  The csvf is deleted.  You should be using the latest source on the `master` branch.

**Comment by fab6** [comment]
hm, good, question, .... I wanted to use https://github.com/firemodels/test_bundles/releases/tag/FDS_TEST

but it seems that I am using 
![image](https://github.com/MaCFP/radi-db/assets/3047695/1fc5b614-3f6c-4bd4-90dc-10425b04df28)

**Comment by rmcdermo** [comment]
Well, should be up to FDS-6.8.0-138-g51097e3-nightly, give that one a try.

**Comment by fab6** [comment]
yes, I am already trying... current status:
![image](https://github.com/MaCFP/radi-db/assets/3047695/3470bbda-7776-49b6-bb02-00bbfce726ab)
but I just gave it 5min, which was maybe to short

**Comment by rmcdermo** [comment]
Fabian,
I think you may be having cluster problems.  This case will actually just run in serial on your desktop machine in less than minute.

**Reply by fab6** [reply]
yes, I will check further. in serial it gives me an error with too many open files. Is there a setting for it which I can use to avoid this too many open file error?

**Reply by rmcdermo** [reply]
This is not an FDS error that I know of.

**Comment by fab6** [comment]
![image](https://github.com/MaCFP/radi-db/assets/3047695/856e57a7-05ea-4c33-834d-46bde04736db)
Ok,  it worked now for temperature on my laptop, I did probably something really clever before

**Reply by rmcdermo** [reply]
OK, great!

**Comment by fab6** [comment]
ok, so next step would be the species... same structure, but all in one file with multiple columns and correct SOOT

**Reply by fab6** [reply]
thanks again!

**Comment by fbraennstroem** [comment]
Hi,
now trying to repeat the temperature initialization I get the mentioned error with "too many open files" on my laptop

![image](https://github.com/MaCFP/radi-db/assets/83709063/e7e99043-ae2f-4af2-a85b-db10fbbe2f08)

This is strange... not sure, what is different now than two weeks ago.
Maybe you have an idea?

**Reply by fbraennstroem** [reply]
here is the rest of the error message, not sure if this is of any help
![image](https://github.com/MaCFP/radi-db/assets/83709063/43b7c800-8605-4a5b-a206-c8ad4580b66b)

Ok, `ulimit -n 2048` helped...

**Comment by fbraennstroem** [comment]
FYI:
Memory usage on my laptop is roughly 6GB for species and temperature for FM Burner case; it takes roughly a 1min 30s

**Comment by fbraennstroem** [comment]
mappings...
![image](https://github.com/MaCFP/radi-db/assets/83709063/46364bc1-72e0-4d1e-adaa-ce167640d090)

**Reply by fbraennstroem** [reply]
and some BNDF snapshots with TEMP and SPECIES read in
![image](https://github.com/MaCFP/radi-db/assets/83709063/59a8585a-687a-4660-8f62-755af19b1dfb)

**Reply by fbraennstroem** [reply]
makes just a bit sense, but not too much...

**Comment by fbraennstroem** [comment]
![image](https://github.com/MaCFP/radi-db/assets/83709063/2900c4b3-6ed3-42e9-9aae-af0f05f59461)
this looks better...

**Comment by fbraennstroem** [comment]
now, I used these settings:
```fortran
&TIME T_END=0.01, DT=0.001/
&DUMP DT_DEVC=0.0000001, DT_SL3D=0.001 ,DT_SLCF=0.001,DT_BNDF=0.001/
&RADI NUMBER_RADIATION_ANGLES=20
      INITIAL_RADIATION_ITERATIONS=10
      TIME_STEP_INCREMENT = 1
      ANGLE_INCREMENT = 1 /
```
There is no visible change in species or temperature during over the simulation time.

The wall is set at 20C
![image](https://github.com/MaCFP/radi-db/assets/83709063/bf66465b-d1d8-4b09-95cb-008440088778)

**Comment by fbraennstroem** [comment]
and 400 angles
![image](https://github.com/MaCFP/radi-db/assets/83709063/ab9140df-12f5-43a1-a969-b77378f6ec7d)
and the top view on the burner
![image](https://github.com/MaCFP/radi-db/assets/83709063/3bf97034-1e76-4d36-97fc-5d218f8d32df)

**Comment by fbraennstroem** [comment]
SOOT: I just used the plain mass fraction; I am not sure at the moment, if we need the aerosol_volume_fraction as soot input or that we would need the soot mass fraction and somehow the local density as well.

**Comment by fbraennstroem** [comment]
Looking at the SUM PRIMITIVE MASS FRACTIONS gives
![image](https://github.com/MaCFP/radi-db/assets/83709063/28480b78-2362-43a3-b27b-6b82d295efe1)
this looks almost too perfect

pressure is not imported, but has only a tiny effect
![image](https://github.com/MaCFP/radi-db/assets/83709063/3eaca956-2ec1-427e-8a53-4df0bd0f8e2d)
![image](https://github.com/MaCFP/radi-db/assets/83709063/e087dbcd-3ab0-4528-b499-7d7dc3926a6f)

**Reply by rmcdermo** [reply]
Mass fractions should sum to unity unless we have a bug.  This gets enforced by absorbing error into the most abundant species.

**Reply by fbraennstroem** [reply]
ah, yes, good point, I think you mentioned this before

**Comment by fbraennstroem** [comment]
and kappa
![image](https://github.com/MaCFP/radi-db/assets/83709063/8af5bbe9-da08-446d-bffb-fc48726e4abf)

**Comment by fbraennstroem** [comment]
I would say that this looks promising...
